### ISSUE TITLE
[gdal:x64-windows-static] missing geos dependencies

### ISSUE THREAD

**Issue by O-Tom** [microsoft/vcpkg#15526]
**Host Environment**
- OS: Windows 10 build 1607
- Compiler: Visual C++ 2017 00370-20001-21499-AA108 / VisualStudio.15.Release/15.9.29+28307.1300

**To Reproduce**
- `.\vcpkg install gdal --triplet=x64-windows-static`
- Write a basic program using gdal and compile/link with cmake.

Linking fails with missing symbols errors from the geos library.
Manually adding the geos package with cmake solves the linking errors (well... almost as it may generate conflicts with windows's `io.h` and `./include/geos/io.h` but that's another story).

Looking at `./ports/gdal/dependency_win.cmake` we see these declarations for geos
```
  # Setup geos libraries + include path
  set(GEOS_INCLUDE_DIR "${CURRENT_INSTALLED_DIR}/include" )
  set(GEOS_LIBRARY_REL "${CURRENT_INSTALLED_DIR}/lib/geos_c.lib" )
  set(GEOS_LIBRARY_DBG "${CURRENT_INSTALLED_DIR}/debug/lib/geos_cd.lib" )
```

But when geos is installed, it comes with two static libs, not one:
![2021-01-08 16_10_05-lib](https://user-images.githubusercontent.com/6274985/104030811-3fc53880-51cc-11eb-955b-2b3bbf494320.png)

So I guess the geos declarations should be modified this way to be complete:
```
  # Setup geos libraries + include path
  set(GEOS_INCLUDE_DIR "${CURRENT_INSTALLED_DIR}/include" )
  set(GEOS_LIBRARY_REL "${CURRENT_INSTALLED_DIR}/lib/geos_c.lib ${CURRENT_INSTALLED_DIR}/lib/geos.lib" )
  set(GEOS_LIBRARY_DBG "${CURRENT_INSTALLED_DIR}/debug/lib/geos_cd.lib ${CURRENT_INSTALLED_DIR}/debug/lib/geosd.lib" )
```

**Comment by JackBoosY** [comment]
GDAL can be built successfully without linking with `geos.lib`/`geosd.lib`, so I don’t think this library is needed when using GDAL. This issue may be caused by the lack of linking to `geos_c.lib`/`geos_cd.lib` when using gdal.

Can you provide the failure logs?

@Neumann-A What do you think about?

**Comment by Neumann-A** [comment]
@JackBoosY: I don't use GDAL itself but I would just link whatever is needed by the buildsystem of GDAL. ()

**Comment by O-Tom** [comment]
I see geos is used throughout OGR when enabled, and OGR is a big chunk of GDAL.
Did you try to link a static version of one of the GDAL tools, say `gdalinfo` ?

In the meantime, I rebuilt the gdal port locally with the suggested modification and got everything working fine.
I can provide failure logs, but I have to restore the gdal port back; it's going to take some time.

**Comment by JackBoosY** [comment]
@O-Tom  Any progress?

**Comment by O-Tom** [comment]
Hi @JackBoosY
Yes and no :)

I have a large project I wish to migrate to vcpkg and compile on both Windows and Linux at the same time.
As I have to move forward and got stuck on gdal more than expected, I simply forked vcpkg and added the geos dependency in my copy, exactly the way I suggested in my initial post. Since then my windows port of GDAL worked quite well.

In the meantime I tried to compile the port under Linux and faced the same problem with geos, along with other problems you are probably aware of (the ./configure script is doing its best to pick system wide installations of libraries and avoid the ones inside vcpkg). For now I'm on plan B which is to `apt-get` proj & gdal on Linux, but I'm really not satisfied with that.

I wish I could provide more help as I will really need the package to be working in both systems in the end.
For now I'm in a rush to complete my initial task and won't get back on those gdal issues in the near future.

**Comment by PhoebeHui** [comment]
@O-Tom, could you try PR https://github.com/microsoft/vcpkg/pull/15292 and see if it solve your problems?

**Comment by O-Tom** [comment]
Sure @PhoebeHui !

I switched to the corresponding repo yesterday and here is what I got from my CI/CD:

The windows port still suffers from missing geos dependencies for the `x64-windows-static` triplet.
Tons of LNK2019 with the same scheme:
```
gdal.lib(geos_ts_c.cpp.obj) : error LNK2019: unresolved external symbol "class std::basic_ostream<char,struct std::char_traits<char> > & __cdecl geos::geom::operator<<(class std::basic_ostream<char,struct std::char_traits<char> > &,class geos::geom::Coordinate const &)" (??6geom@geos@@YAAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@AEAV23@AEBVCoordinate@01@@Z) referenced in function "public: char * __cdecl <lambda_ecc4d6888735fe1fa3510953ded65e07>::operator()(void)const " (??R<lambda_ecc4d6888735fe1fa3510953ded65e07>@@QEBAPEADXZ)
```
Manually adding the geos dependency to the build files or patching vcpkg as suggested in my initial post solves this issue.

The linux port fails at install time (`x64-linux` triplet) with a compilation error.
Here is the full content of the `build-x64-linux-dbg-err.log` file.
```
In file included from ogrogdidriver.cpp:30:
ogrogdi.h:40:10: fatal error: ecs.h: Aucun fichier ou dossier de ce type
   40 | #include "ecs.h"
      |          ^~~~~~~
compilation terminated.
make[3]: *** [../../../GDALmake.opt:646 : ../o/ogrogdidriver.lo] Erreur 1
make[2]: *** [../../GDALmake.opt:659 : ogdi-target] Erreur 2
make[2]: *** Attente des tâches non terminées....
make[1]: *** [GNUmakefile:58 : sublibs] Erreur 2
make: *** [GNUmakefile:100 : ogr-target] Erreur 2
```

**Comment by JackBoosY** [comment]
This issue will be fixed by https://github.com/microsoft/vcpkg/pull/16088 or #15808.

**Comment by dg0yt** [comment]
I'm currently trying to improve the Unix part of the gdal portfile. 

> In the meantime I tried to compile the port under Linux and faced the same problem with geos, along with other problems you are probably aware of (the ./configure script is doing its best to pick system wide installations of libraries and avoid the ones inside vcpkg).

For geos and static linking on linux, I found that the geos C library depends on the geos C++ library which depends on the C++ runtime. However, during the `configure` tests, the C++ runtime is not linked so that test fails due to unresolved symbols... which seems similar to the error reported for x64-windows-static (`unresolved external symbol "class std::basic_ostream`).

Hidden dependencies are a reason for more configure test failures on Linux, partly hidden by how `LIBS` is initialized in the port file.

**Comment by FrankXie05** [comment]
@O-Tom The resolved PR #20502 has been merged, can you use the latest vcpkg library and try again? :)

**Comment by JackBoosY** [comment]
We don't have enough information to solve this issue, unfortunately; if you have any more information to help us solve this, please reopen!
